guide_service_errors.md

```diff
@@ -0,0 +1,234 @@
+  INVALID_PERMISSIONS(HttpStatus.S_403_FORBIDDEN, "User does not have valid permissions", null),
+  ILLEGAL_PARAM(HttpStatus.S_400_BAD_REQUEST, "Request parameter cannot be used", null);
+  // Add more service errors here...
+
+  SampleServiceError(HttpStatus status, String message, Class<? extends RecordTemplate> errorDetailType)
+  {
+    _httpStatus = status;
+    _message = message;
+    _errorDetailType = errorDetailType;
+  }
+
+  private final HttpStatus _status;
+  private final String _message;
+  private final Class<? extends RecordTemplate> _errorDetailType;
+
+  /*
+   * Suggested: Define string constants for each of the error codes for easy access.
+   * These codes should match the enum value names.
+   */
+  public class Codes
+  {
+    QUOTA_EXCEEDED = "QUOTA_EXCEEDED";
+    INVALID_PERMISSIONS = "INVALID_PERMISSIONS";
+  }
+
+  @Override
+  public HttpStatus httpStatus()
+  {
+    return _status;
+  }
+
+  @Override
+  public String code()
+  {
+    return name();
+  }
+
+  @Override
+  public String message()
+  {
+    return _message;
+  }
+
+  @Override
+  public Class<? extends RecordTemplate> errorDetailType()
+  {
+    return _errorDetailType;
+  }
+}
+```
+
+## Step 3: Apply the Service Errors to your Resource
+
+Now that you have defined your service errors, you can use them to specify which resources
+or particular resource methods may return such errors. You can do this by using a few
+Java annotations described below.
+
+### `@ServiceErrorDef`
+
+In order to apply your service error definitions to a particular resource, you first need
+to use the `@ServiceErrorDef` annotation at the resource level to reference your enum.
+This class-level annotation is used to reference your service error enum definition.
+This is used by Rest.li during build-time when generating IDLs and also during runtime
+when validating service errors returned by your resource. Without this annotation,
+Rest.li will have no way of referencing the information in your service error definition.
+
+```java
+@ServiceErrorDef(SampleServiceError.class)
+class MyResource extends CollectionResourceTemplate<Long, MyRecord>
+{
+  // Resource methods here...
+}
+```
+
+### `@ServiceErrors`
+
+Next, you can use the `@ServiceErrors` annotation to
+indicate which service error codes may be returned by a
+particular resource or method. It can be used as a class-level annotation (to refer
+to a resource) or as a method-level annotation (to refer to a resource method).
+This is used during build-time when generating IDLs and also during runtime when
+validating service errors returned by your resource.
+In the following example,
+all methods in the resource may return a `QUOTA_EXCEEDED` error, but only `getAll` can return
+an `INVALID_PERMISSIONS` error:
+
+```java
+@ServiceErrorDef(SampleServiceError.class)
+@ServiceErrors(SampleServiceError.Codes.QUOTA_EXCEEDED)
```

Comment:
Won't it be easier for the developer to write something like
`@ServiceErrors(SampleServiceError.QUOTA_EXCEEDED)`
`QUOTA_EXCEEDED` would be a `ServiceError` type. At the build time, the framework can use it to generate the IDL and at the execution to validate. 
The benefit is to have to set only 1 annotation, not 2.

Reply:
This is something Gevorg, Karthik, and I discussed, but it's not possible. Java annotations only allow certain return types in their methods: https://docs.oracle.com/javase/specs/jls/se8/html/jls-9.html#jls-9.6.1

Reply:
Java does allow enums in this case, but we can't use interfaces that enums implement, nor can we use a generic enum class to capture all possible enums.

Reply:
Thanks for the explanation. I remember that now.